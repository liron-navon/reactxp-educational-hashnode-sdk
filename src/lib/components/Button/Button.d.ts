/// <reference types="react" />
import { Types, Animated, Component } from 'reactxp';
interface ButtonProps extends Types.ButtonProps {
    /** the type of the button */
    type?: 'default' | 'primary' | 'secondary' | 'flat';
    /** extra style for the button title */
    titleStyle?: Types.StyleRuleSet<Types.TextStyle>;
    /** uppercase the title */
    upper?: boolean;
    /** disable the hover style, hover style only work on the web */
    disableHoverStyle?: boolean;
}
/** A simple button component */
declare class Button extends Component<ButtonProps> {
    state: {
        isPressed: boolean;
    };
    hoverAnimation: Types.Animated.CompositeAnimation | null;
    animatedOpacity: Animated.Value;
    animateHover(on: boolean): void;
    setPressed(isPressed: boolean): void;
    render(): JSX.Element;
}
export default Button;
