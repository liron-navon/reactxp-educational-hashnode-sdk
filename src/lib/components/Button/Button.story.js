import React from 'react';
import {storiesOf} from '@storybook/react';
import Button from './Button';
import {action} from '@storybook/addon-actions';

storiesOf('components|Button', module)
    .add('default', () => {
        return (
            <Button
                accessibilityLabel={'default button'}
                onPress={action('onPress')}
                title={'Default Button'}
                upper
            />
        )
    })
    .add('primary', () => {
        return (
            <Button
                type={'primary'}
                accessibilityLabel={'primary button'}
                onPress={action('onPress')}
                title={'Primary Button'}
                upper
            />
        )
    })
    .add('secondary', () => {
        return (
            <Button
                type={'secondary'}
                accessibilityLabel={'secondary button'}
                onPress={action('onPress')}
                title={'Secondary Button'}
                upper
            />
        )
    })
    .add('flat', () => {
        return (
            <Button
                type={'flat'}
                accessibilityLabel={'flat button'}
                onPress={action('onPress')}
                title={'Flat Button'}
                upper
            />
        )
    })
    .add('mobile style',
        () => {
            return (
                <div>
                    <Button
                        accessibilityLabel={'default button'}
                        onPress={action('onPress')}
                        title={'Default Button'}
                        upper
                        disableHoverStyle
                    />
                    <Button
                        type={'primary'}
                        accessibilityLabel={'primary button'}
                        onPress={action('onPress')}
                        title={'Flat Button'}
                        upper
                        disableHoverStyle
                    />
                    <Button
                        type={'secondary'}
                        accessibilityLabel={'secondary button'}
                        onPress={action('onPress')}
                        title={'Flat Button'}
                        upper
                        disableHoverStyle
                    />
                </div>
            )
        },
        {info: {components: {Button}}});
