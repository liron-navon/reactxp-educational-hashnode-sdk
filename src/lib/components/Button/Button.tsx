import React from 'react';
import {
    Button as RXButton,
    Text,
    Styles,
    Types,
    Animated,
    Component
} from 'reactxp';
import {createStyleSheet, optionalStyle} from '../../utils/styleHelpers';
import theme from "../../theming/theme";
import {isReactNative} from "../../utils/platform";

// the prop types for the button
interface ButtonProps extends Types.ButtonProps {
    /** the type of the button */
    type?: 'default' | 'primary' | 'secondary' | 'flat',
    /** extra style for the button title */
    titleStyle?: Types.StyleRuleSet<Types.TextStyle>,
    /** uppercase the title */
    upper?: boolean,
    /** disable the hover style, hover style only work on the web */
    disableHoverStyle?: boolean
}

function createButtonStyles() {
    const styles = createStyleSheet({
        root: {
            padding: 0,
            margin: 0
        },
        buttonBase: {
            margin: 16,
            borderRadius: 4,
            ...theme.shadow(1)
        },
        buttonPressed: {
            ...theme.shadow(2)
        },
        text: {
            fontSize: 16,
            marginVertical: theme.spacing,
            marginHorizontal: theme.spacing * 2,
            textAlign: 'center'
        }
    });

// add some extra styling for different button types
    const buttonTypeStyles = {
        default: createStyleSheet({
            button: {
                backgroundColor: theme.colors.grey,
            },
            text: {
                color: theme.colors.text
            }
        }),
        primary: createStyleSheet({
            button: {
                backgroundColor: theme.colors.primary,
            },
            text: {
                color: theme.colors.textContrast
            }
        }),
        secondary: createStyleSheet({
            button: {
                backgroundColor: theme.colors.secondary,
            },
            text: {
                color: theme.colors.textContrast
            }
        }),
        flat: createStyleSheet({
            button: {
                backgroundColor: theme.colors.transparent,
                ...theme.shadow(0)
            },
            text: {
                color: theme.colors.text
            }
        })
    };

    return {styles, buttonTypeStyles};
}

/** A simple button component */
class Button extends Component<ButtonProps> {
    state = {
        isPressed: false
    };

    // the opacity animation to allow better control of it
    hoverAnimation: Types.Animated.CompositeAnimation | null = null;

    // this value is used to animate the opacity
    animatedOpacity: Animated.Value = Animated.createValue(1);

    // apply hover animation to the button
    animateHover(on: boolean): void {
        if (this.hoverAnimation) {
            this.hoverAnimation.stop();
        }
        this.hoverAnimation = Animated.timing(this.animatedOpacity,
            {
                toValue: on ? 0.5 : 1,
                duration: 320,
                easing: Animated.Easing.InOut()
            }
        );
        this.hoverAnimation.start();
    }

    setPressed(isPressed: boolean): void {
        this.setState({
            isPressed
        })
    }

    render() {
        // separate the props we handle manually and the native rx props
        const {
            type, titleStyle, upper,
            disableHoverStyle, onHoverStart,
            onHoverEnd, onPressIn, onPressOut,
            style, title, ...buttonProps
        } = this.props;
        const {isPressed} = this.state;
        const {styles, buttonTypeStyles} = createButtonStyles();

        // set the button type
        const typeStyle = buttonTypeStyles[type || 'default'] || buttonTypeStyles.default;

        // create animated style
        const animatedButtonStyle = isReactNative() || isPressed
            ? {}
            : Styles.createAnimatedViewStyle({
                opacity: this.animatedOpacity,
            });

        return (
            <RXButton
                {...buttonProps}
                style={styles.root}
                onHoverStart={(...event) => {
                    !disableHoverStyle && this.animateHover(true);
                    onHoverStart && onHoverStart(...event);
                }}
                onHoverEnd={(...event) => {
                    !disableHoverStyle && this.animateHover(false);
                    onHoverEnd && onHoverEnd(...event);
                }}
                onPressIn={(...event) => {
                    this.setPressed(true);
                    onPressIn && onPressIn(...event);
                }}
                onPressOut={(...event) => {
                    this.setPressed(false);
                    onPressOut && onPressOut(...event);
                }}>
                <Animated.View
                    style={[
                        styles.buttonBase,
                        optionalStyle(isPressed, styles.buttonPressed),
                        typeStyle.button,
                        animatedButtonStyle,
                        style
                    ]}>
                    <Text style={[
                        styles.text,
                        typeStyle.text,
                        titleStyle
                    ]}>
                        {
                            upper
                                ? title!.toUpperCase()
                                : title
                        }
                    </Text>
                </Animated.View>
            </RXButton>
        )
    }
}

export default Button;
