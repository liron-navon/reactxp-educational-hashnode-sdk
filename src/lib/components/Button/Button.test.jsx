import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button';

test('renders correctly', () => {
    const tree = renderer.create(<Button/>).toJSON();
    expect(tree).toMatchSnapshot();
});
