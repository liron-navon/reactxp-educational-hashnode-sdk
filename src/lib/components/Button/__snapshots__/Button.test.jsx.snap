// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  role="button"
  style={
    Object {
      "alignItems": "stretch",
      "appRegion": "no-drag",
      "backgroundColor": "#7d88a9",
      "borderColor": "transparent",
      "borderRadius": 16,
      "borderStyle": "solid",
      "borderWidth": "0",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "column",
      "flexGrow": 0,
      "flexShrink": 0,
      "justifyContent": "center",
      "margin": 16,
      "overflow": "hidden",
      "padding": 0,
      "position": "relative",
      "textAlign": "left",
    }
  }
>
  <div
    style={
      Object {
        "color": "white",
        "cursor": "inherit",
        "display": "inline",
        "flexGrow": 0,
        "flexShrink": 0,
        "fontSize": 16,
        "marginBottom": 6,
        "marginLeft": 12,
        "marginRight": 12,
        "marginTop": 6,
        "overflow": "hidden",
        "overflowWrap": "break-word",
        "position": "relative",
        "whiteSpace": "pre-wrap",
      }
    }
  />
</button>
`;
